libcob/intrinsic.c

```diff
@@ -3153,97 +3154,178 @@ cob_switch_value (const int id)
 void
 cob_decimal_pow (cob_decimal *pd1, cob_decimal *pd2)
 {
-	cob_uli_t		n;
-	const int		sign = mpz_sgn (pd1->value);
+	int sign_nbr;
+	int sign_exp;
+	int power_case;
+	int negat_result = 0 ;
 
-	if (unlikely (pd1->scale == COB_DECIMAL_NAN)) {
+	if (unlikely(pd1->scale == COB_DECIMAL_NAN)) {
 		return;
 	}
-	if (unlikely (pd2->scale == COB_DECIMAL_NAN)) {
+	if (unlikely(pd2->scale == COB_DECIMAL_NAN)) {
 		pd1->scale = COB_DECIMAL_NAN;
 		return;
 	}
 
-	if (mpz_sgn (pd2->value) == 0) {
-		/* Exponent is zero */
-		if (sign == 0) {
-			/* 0 ^ 0 */
-			cob_set_exception (COB_EC_SIZE_EXPONENTIATION);
+	sign_nbr = mpz_sgn(pd1->value);
+	sign_exp = mpz_sgn(pd2->value);
+
+	power_case = sign_nbr * sign_exp;
+
+	cob_trim_decimal(pd2);
+	cob_trim_decimal(pd1);
+
+	switch (power_case) {
+	
+	case  1:
+	case -1:
+		if (pd2->scale != 0 && sign_nbr == -1) {
+			/* Case number < 0 and decimal exponent --> Error */
+			pd1->scale = COB_DECIMAL_NAN;
+			cob_set_exception(COB_EC_SIZE_EXPONENTIATION);
+			return;
 		}
-		mpz_set_ui (pd1->value, 1UL);
-		pd1->scale = 0;
-		return;
+		goto Compute_Power;
```

Comment:
please reorder that case to have zero first, this can then `return` while the fall-through of 1/-1 is easier to follow and we don't need the `Compute_Power` label any more